### ISSUE TITLE
OOM issue when training 1 billion corpus

### ISSUE THREAD

**Issue by deep-speech** [kimiyoung/transformer-xl#52]
I am trying to train with 1 billion corpus on Tesla P40.  Following are the values being used.

N_LAYER = 12
D_MODEL = 512, 
D_EMBED = 512, 
D_INNER = 2048,
D_HEAD = 64

I also tried with a BSZ of 128, it still gives OOM error.

**Comment by kimiyoung** [comment]
Did you try these hyperparams? https://github.com/kimiyoung/transformer-xl/blob/master/tf/scripts/lm1b_base_gpu.sh

If it's still OOM, then I guess there's not much we can do except for further decreasing the model size and batch size.

**Comment by deep-speech** [comment]
I am able to train with N_LAYERS=16 and BSZ=32,  BSZ=40 gave OOM. Rest of the hyperparams are same as lm1b_base_gpu.sh. 

What will be ideal set of hyperparams for training a very large corpus(~20 billion words) ?

**Comment by kimiyoung** [comment]
For 20 billion words, it is necessary to use subword units as inputs, such as BPEs. And then use the largest model size that fits your hardware.

**Comment by deep-speech** [comment]
Model has been training on 1 billion corpus for 1 week with 8 GPUs, N_LAYERS=16 and BSZ=32, all other hyperparams same as lm1b_base_gpu.sh, training logs show something like - loss 7.20 | pplx 1342.52, bpc 10.3907. I am not sure if it is really converging at all.

**Comment by kimiyoung** [comment]
It is not converging. In fact, it's not being optimized at all... This is because the batch size is too small. You will need to increase the batch size, decrease the learning rate, or use a larger number of warmup steps.

**Comment by deep-speech** [comment]
Model is converging well, I have one query though - why is --dataset=enwik8 for elif [[ $1 == 'test_data' ]] in lm1b_base_gpu.sh ?

**Comment by cmathx** [comment]
@deep-speech I trained 1 billion corpus using the same configuration(N_LAYERS=16 and BSZ=32) with 8 GPUs, how much time it will converge? Did you adjust other hyper parameters?
